### DISCUSSION TITLE
[Now] Best way to set the Cache-Control header for an API Route?

### DISCUSSION THREAD

**Discussion by Danetag** [vercel/next.js#10729] [Help] [answered]
Hey people!

I've been trying to set the `cache-control` to `s-maxage=60, stale-while-revalidate` for the API Routes on a NextJS app deployed on Now.

I tried 2 ways:
- setting the header before sending the data, directly in the API Route function.

```.js
...
res.setHeader('Cache-Control', 's-maxage=60, stale-while-revalidate');
res.status(status).json(data);
```

- setting rules in `now.json`.

```.json
...
"routes": [
    {
      "src": "/api/(.*)",
      "headers": {
        "cache-control": "s-maxage=60, stale-while-revalidate"
      },
      "dest": "/api/$1",
      "continue": true
    }
  ]
```

...None of them work once deployed :/ What would be the best way?

Thanks!

**Comment by timneutkens** [comment, marked as answer]
The first method is what you'll want to do, and it seems to work fine:

```js
// pages/api/people/index.js
import { people } from '../../../data'

export default (req, res) => {
  res.setHeader('Cache-Control', 's-maxage=60, stale-while-revalidate');
  res.status(200).json(people)
}
```

```
curl -I https://api-routes-app-timzeit.zeit.now.sh/api/people
```

First request is expectedly a `MISS` as it's the first request:

```
curl -I https://api-routes-app-timzeit.zeit.now.sh/api/people
HTTP/2 200 
content-type: application/json; charset=utf-8
cache-control: public, max-age=0, must-revalidate
date: Fri, 28 Feb 2020 09:37:01 GMT
content-length: 0
x-now-cache: MISS
x-now-trace: bru1
server: now
strict-transport-security: max-age=63072000; includeSubDomains; preload
```

Second request is a `HIT`:

```
curl -I https://api-routes-app-timzeit.zeit.now.sh/api/people
HTTP/2 200 
content-type: application/json; charset=utf-8
cache-control: public, max-age=0, must-revalidate
date: Fri, 28 Feb 2020 09:37:01 GMT
content-length: 0
x-now-cache: HIT
x-now-trace: bru1
server: now
strict-transport-security: max-age=63072000; includeSubDomains; preload
```


Note that if the status is an error code (404, 500, etc) the result won't be cached

**Reply by trigunam** [reply]
@timneutkens We have around 25+ api/* implementations and for each res we need to add this logic. is there a better way to improve this to avoid duplicates and also it would be missed by some developers if it's not standardized?

**Comment by Danetag** [comment]
@timneutkens Ohh ok, if this is the way to make sure it works, then it's all good. 
I was focusing on the `cache-control` response `public, max-age=0, must-revalidate`, but I do notice the same behavior than you mentionned: a `MISS`, then a `HIT`. My bad!

Thank you!

**Reply by timneutkens** [reply]
Yeah that's how it's supposed to work as the `cache-control` header returned from the server ensures to not get stale data because of other proxies.

- `s-maxage` makes sure the response is cached on the proxy
- `stale-while-revalidate` ensures that after the first request (the MISS) the response users get is not a MISS but a stale (cached) version, then in the background the function gets called and the cache is updated for future requests.

**Reply by jasondonnette** [reply]
I'm also curious if we can modify the `public, max-age=0, must-revalidate` behavior on the cached response. I have a function set up that serves images, and I think this means that my browser re-downloads the image each time it's used instead of grabbing it from memory. 

So when navigating around on an app, I am getting a bunch of network requests from images that I would want to cache locally.

Should I be setting up an additional CDN layer (cloudfront, etc) if I'm using functions to serve static content like images? 

Or am I misunderstanding which header(s) to set to get this behavior?

**Reply by jasondonnette** [reply]
Just answered my own question, putting it here in case any stumbles across this wondering the same thing:

`max-age` is the header I was looking for. So to get the client-side caching as well, a cache-control header of something like:

`max-age=604800, s-maxage=604800, stale-while-revalidate` will do the trick.

**Reply by timneutkens** [reply]
Overriding the default cache-control header is indeed possible by additionally setting the browser values next to `s-maxage` 👍

**Reply by cdock1029** [reply]
@timneutkens I'm a little new to caching. I understand the `HIT` / `MISS` dynamic. But in my api I'm sending a token in the Authorization header. I'm working through [this article ](https://sookocheff.com/post/api/effective-caching/#handling-authentication)to understand better, I but I'm not quite sure what exactly the `BYPASS` value I'm seeing means, aside from the short explanation in the docs.

Are my requests going all the way  to origin server or not? Are requests with authorization cache-able or not?

Would adding `public` only apply for the browser? Not sure what is best practice here

**Reply by JannikZed** [reply]
We are facing the same issue with the authorization header. We want to send this header to our API, but still use the caching on the edge.. is there any possibility to realize that?

**Comment by trigunam** [comment]
@timneutkens Is there an improved answer based on [Custom Headers](https://nextjs.org/docs/api-reference/next.config.js/headers) in next.conf.js in the Next 9.5 & up release?

Why is this not an acceptable answer?

`"headers": [{
   "key": "Cache-Control"
   "value": "no-cache no-store"
}]`

**Reply by grrowl** [reply]
from that page,

> Cache-Control headers set in next.config.js will be overwritten in production to ensure that static assets can be cached effectively

**Comment by Ali-Hussein-dev** [comment]
I am new to cache control, and it seems to me weird why is this ?!! 
when the req is "Hit"  `cache-control: public, max-age=0, must-revalidate` and not `cache-control: public, s-maxage=60, stale-while-revalidate`.

**Comment by jflayhart** [comment]
We're finding on our end that the request header `cache-control` is one of `max-age: 0` OR `no-cache` and therefore is causing our CDN to bypass the cache and go straight to origin (nextjs). Would using a custom server have anything to do with those request headers (we are not setting them anywhere)? We want them to be `public, max-age=1234` for example, but never seems to work for SSG pages and HTML.

On the other hand, all `/static` assets are cached by our CDN. Just find it odd we can't cache HTML? How can we override the cache-control for HTML pages (ie SSG)?
